File: core/bindings/python/auki-sdk-py/examples/robot_task.py
"""Run a provisioned robot; its optional peer remains connected between tasks.

Live operation: registers /example/inspect-input/v1, claims work, reads input_id
and writes a report. The submitter owns result cleanup. No hardware is controlled.
"""
import asyncio
import os
import signal

import auki_sdk

CAPABILITY = "/example/inspect-input/v1"


async def inspect_input(task):
    data = task.data()
    await task.progress({"phase": "reading"})
    source = await data.read(task.meta["input_id"])
    peer = task.peer()
    # This optional operation requires a build with the Info protocol enabled.
    # The remote application remains responsible for authorizing its requests.
    if task.meta.get("remote_peer_id"):
        if peer is None:
            raise ValueError("this task requires configured P2P")
        client = auki_sdk.AukiInfoClient(peer)
        await client.fetch_exact(task.meta["remote_peer_id"], task.meta["remote_route"])
    stored = await data.write(f"Read {len(source)} bytes".encode(),
        name=f"inspection-{task.id}", data_type="example.report.v1")
    return {"output_cids": [stored["id"]]}


async def main():
    robot = auki_sdk.AukiRobotCredential(
        dds_url=os.environ["DDS_BASE_URL"], dms_url=os.environ["DMS_BASE_URL"],
        registration=os.environ["ROBOT_REGISTRATION_CREDENTIAL"],
        audience=os.environ.get("DDS_ROBOT_AUDIENCE"),
        version="1.0.0", client_id=os.environ["AUKI_CLIENT_ID"], capabilities=[CAPABILITY],
        peer_identity_file=os.environ.get("AUKI_PEER_IDENTITY_FILE"),
    )
    tasks = auki_sdk.AukiDmsTasks(robot, {CAPABILITY: inspect_input})
    async def serve():
        await tasks.start()
        peer = tasks.peer()
        if peer is not None:
            print(f"Robot peer ready: {peer.peer_id}")
        # Process-wide endpoints may use this peer while idle. Close them before
        # tasks.close(); task-specific endpoints belong in the handler's finally.
        await tasks.run()
    running = asyncio.ensure_future(serve())
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, running.cancel)
    try:
        await running
    except asyncio.CancelledError:
        pass
    finally:
        try:
            await tasks.close()
        finally:
            await robot.close()
            for sig in (signal.SIGINT, signal.SIGTERM):
                loop.remove_signal_handler(sig)


if __name__ == "__main__":
    asyncio.run(main())

File: core/auki-sdk/examples/robot_task.rs
//! Run one task as a provisioned robot. No hardware is controlled.
//! Live operation: registers, reads task data and reports a result to DMS.
use async_trait::async_trait;
use auki_sdk::{
    AukiDmsTasks, AukiPeerConfig, AukiRobotCredential, AukiTaskPeerConfig, Identity, RobotConfig,
    SecretString, TaskContext, TaskError, TaskHandler, TaskPeerContext, TaskResult, TasksConfig,
};
use std::sync::Arc;
use tokio_util::sync::CancellationToken;

const CAPABILITY: &str = "/example/inspect-input/v1";
struct Inspect;
#[async_trait]
impl TaskHandler for Inspect {
    async fn run(&self, task: TaskContext) -> Result<TaskResult, TaskError> {
        let id = task.task.meta["input_id"]
            .as_str()
            .ok_or(TaskError::Handler)?
            .parse()
            .map_err(|_| TaskError::Handler)?;
        task.progress(serde_json::json!({"phase": "reading"}))?;
        let bytes = task.data().read(id).await.map_err(|_| TaskError::Handler)?;
        let peer_id = task.peer().map(|peer| peer.peer_id().to_string());
        Ok(TaskResult {
            output_cids: vec![],
            meta: serde_json::json!({"bytes": bytes.len(), "peer_id": peer_id}),
        })
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let dds = std::env::var("DDS_BASE_URL")?;
    let dms = std::env::var("DMS_BASE_URL")?;
    let audience = match std::env::var("DDS_ROBOT_AUDIENCE") {
        Ok(value) => Some(value),
        Err(std::env::VarError::NotPresent) => None,
        Err(error) => return Err(error.into()),
    };
    let capabilities = vec![CAPABILITY.into()];
    let mut config = RobotConfig::new(
        &dds,
        &dms,
        SecretString::new(std::env::var("ROBOT_REGISTRATION_CREDENTIAL")?),
        "1.0.0",
        &std::env::var("AUKI_CLIENT_ID")?,
        audience.as_deref(),
        capabilities.clone(),
    )?;
    let peer = std::env::var_os("AUKI_PEER_IDENTITY_FILE")
        .map(|file| {
            AukiTaskPeerConfig::new(
                Identity::load_or_create(file)
                    .map_err(|_| TaskError::Configuration("cannot load peer identity"))?,
                &dds,
                AukiPeerConfig::new(&dms)
                    .map_err(|_| TaskError::Configuration("invalid peer configuration"))?,
            )
        })
        .transpose()?;
    config.peer_identity = peer.as_ref().map(AukiTaskPeerConfig::identity_proof);
    let robot = AukiRobotCredential::new(config)?;
    let tasks = match peer {
        Some(peer) => AukiDmsTasks::new_with_peer(
            robot.clone(),
            capabilities,
            TasksConfig::default(),
            Arc::new(peer),
        )?,
        None => AukiDmsTasks::new(robot.clone(), capabilities, TasksConfig::default())?,
    };
    let cancellation = CancellationToken::new();
    let outcome = async {
        tasks.start(&cancellation).await?;
        if let Some(peer) = tasks.peer() {
            println!("Robot peer ready: {}", peer.peer_id());
        }
        // The same peer stays connected after this task, until tasks.close().
        tasks.run_once(CAPABILITY, &Inspect, &cancellation).await
    }
    .await;
    let cleanup = tasks.close().await;
    robot.close().await;
    cleanup?;
    println!("{:?}", outcome?);
    Ok(())
}
